pyo3-macros-backend/src/method.rs

```diff
@@ -604,107 +603,123 @@ impl<'a> FnSpec<'a> {
+                    bail_spanned!(nv.eq_token.span() => format!("`#[{}]` does not take any arguments", ident))
                 }
             }
-            syn::Meta::List(ref ml @ syn::MetaList { ref path, .. }) => {
-                if path.is_ident("new") {
-                    set_ty!(MethodTypeAttribute::New, path);
-                } else if path.is_ident("setter") || path.is_ident("getter") {
-                    if let syn::AttrStyle::Inner(_) = attr.style {
-                        bail_spanned!(
-                            attr.span() => "inner attribute is not supported for setter and getter"
-                        );
-                    }
-
-                    if path.is_ident("setter") {
-                        set_ty!(MethodTypeAttribute::Setter, path);
-                    } else {
-                        set_ty!(MethodTypeAttribute::Getter, path);
-                    };
-
-                    ensure_spanned!(
-                        python_name.is_none(),
-                        python_name.span() => "`name` may only be specified once"
-                    );
+        }
 
-                    if let Ok(ident) = ml.parse_args::<syn::Ident>() {
-                        python_name = Some(ident);
-                    } else if let Ok(syn::Lit::Str(s)) = ml.parse_args::<syn::Lit>() {
-                        python_name = Some(s.parse()?);
+        fn extract_name(meta: &syn::Meta, ident: &str) -> Result<Option<Ident>> {
+            match meta {
+                syn::Meta::Path(_) => Ok(None),
+                syn::Meta::NameValue(nv) => bail_spanned!(
+                    nv.eq_token.span() => format!("expected `#[{}(name)]` to set the name", ident)
+                ),
+                syn::Meta::List(ml) => {
+                    if let Ok(name) = ml.parse_args::<syn::Ident>() {
+                        Ok(Some(name))
+                    } else if let Ok(name) = ml.parse_args::<syn::LitStr>() {
+                        name.parse().map(Some)
                     } else {
-                        return Err(syn::Error::new_spanned(
-                            ml,
-                            "expected ident or string literal for property name",
-                        ));
+                        bail_spanned!(ml.tokens.span() => "expected ident or string literal for property name");
                     }
-                } else {
-                    new_attrs.push(attr)
                 }
             }
-            syn::Meta::NameValue(_) => new_attrs.push(attr),
+        }
+
+        let meta = &attr.meta;
+        let path = meta.path();
+
+        if path.is_ident("new") {
+            ensure_no_arguments(meta, "new")?;
+            Ok(Some(MethodTypeAttribute::New(path.span())))
+        } else if path.is_ident("__new__") {
+            // TODO deprecate this form?
+            ensure_no_arguments(meta, "__new__")?;
+            Ok(Some(MethodTypeAttribute::New(path.span())))
```

Comment:
I discovered `#[__new__]` is a valid attribute in our code, I think as this is undocumented and redundant with `#[new]` we should deprecate it.